README.md
# skinner: aggregate multi-dimensional data points

Skinner provides functions for summarizing a bunch of multi-dimensional data
points.  It's best described by example.  Here are some data points describing
populations of US cities:

```javascript
datapoints = [
    { 'fields': { 'city': 'Springfield', 'state': 'MA' }, 'value': 153000 },
    { 'fields': { 'city': 'Boston',      'state': 'MA' }, 'value': 636000 },
    { 'fields': { 'city': 'Worcestor',   'state': 'MA' }, 'value': 183000 },
    { 'fields': { 'city': 'Fresno',      'state': 'CA' }, 'value': 505000 },
    { 'fields': { 'city': 'Springfield', 'state': 'OR' }, 'value':  60000 },
    { 'fields': { 'city': 'Portland',    'state': 'OR' }, 'value': 600000 }
];
```

You can sum all of the values:

```javascript
assert.deepEqual(skinner.aggregate(datapoints), [ 2137000 ]);
```

In this case, `aggregate` returned an array of one aggregated data point, which
was just the number `2137000`.

You can break out the results by 'state':

```javascript
assert.deepEqual(skinner.aggregate(datapoints, [ 'state' ]),
    [ [ 'MA', 972000 ],
      [ 'CA', 505000 ],
      [ 'OR', 660000 ] ]);
```

In this case, `aggregate` summed the values and grouped the sums by state.

Obviously, you can do the same with city name (*not* including state), which
sums results of Springfield, since that's in two states:

```javascript
assert.deepEqual(skinner.aggregate(datapoints, [ 'city' ]),
    [ [ 'Springfield', 213000 ],
      [ 'Boston',      636000 ],
      [ 'Worcestor',   183000 ],
      [ 'Fresno',      505000 ],
      [ 'Portland',    600000 ] ]);
```

You can also break out the results by more than one column:

```javascript
assert.deepEqual(skinner.aggregate(datapoints, [ 'state', 'city' ]),
    [ [ 'MA', 'Springfield', 153000 ],
      [ 'MA', 'Boston',      636000 ],
      [ 'MA', 'Worcestor',   183000 ],
      [ 'CA', 'Fresno',      505000 ],
      [ 'OR', 'Springfield', 60000 ],
      [ 'OR', 'Portland',    600000 ] ]);

```

The order that you specify breakdowns determines the order in which the values
are output.  If you do `[ 'city', 'state' ]`, you'll get data points that look
like `[ 'Springfield', 'MA', 153000 ]`.  If you do `[ 'state', 'city' ]`, you'll
get `[ 'MA', 'Springfield', 153000 ]`.



## Bucketizing numbers

Here's a set of data points describing CPU utilization on four 2-CPU systems
where the CPUs are called "cpu0" and "cpu1" on each system:

```javascript
datapoints = [
    { 'fields': { 'host': 'host1', 'cpu': 'cpu0', 'util': 83 }, 'value': 1 },
    { 'fields': { 'host': 'host1', 'cpu': 'cpu1', 'util': 13 }, 'value': 1 },
    { 'fields': { 'host': 'host2', 'cpu': 'cpu0', 'util': 37 }, 'value': 1 },
    { 'fields': { 'host': 'host2', 'cpu': 'cpu1', 'util': 53 }, 'value': 1 },
    { 'fields': { 'host': 'host3', 'cpu': 'cpu0', 'util': 88 }, 'value': 1 },
    { 'fields': { 'host': 'host3', 'cpu': 'cpu1', 'util':  9 }, 'value': 1 },
    { 'fields': { 'host': 'host4', 'cpu': 'cpu0', 'util': 98 }, 'value': 1 },
    { 'fields': { 'host': 'host4', 'cpu': 'cpu1', 'util':  5 }, 'value': 1 }
];
```

As above, the "value" here is the *count* of data points having the specified
fields (e.g., the count of CPUs on host "host1" with name "cpu0" and utilization
"83", which is always 1 in this case).

The difference between this data set and the previous one is that "util" is an
arbitrary number, and if we want to break out results by "util", we probably
don't want a separate result for every possible utilization value.  Instead, we
*bucketize* the utilization values.  We'll use a linear bucketizer with step
"10", which means we'll group the "util" values into equal-sized buckets of 10
units each:

```javascript
bucketizers = {
    'util': skinner.makeLinearBucketizer(10)
};
```

Now, you can summarize CPU utilization across all systems with a single
histogram:

```javascript
var expand = mod_skinner.ordinalToBounds.bind(null, bucketizers.util);
assert.deepEqual(
    expand(skinner.aggregate(datapoints, [ 'util' ], bucketizers)),
    [ [ [ [ 0, 9 ], 2 ],
        [ [ 10, 19 ], 1 ],
        [ [ 30, 39 ], 1 ],
        [ [ 50, 59 ], 1 ],
        [ [ 80, 89 ], 2 ],
        [ [ 90, 99 ], 1 ] ] ]);
```

These results tell us that we had 2 CPUs with utilization between 0 and 9
(inclusive), 2 CPUs with utilization between 80 and 89, and 1 CPU each having
utilizations 10-19, 30-39, and 90-99.

If you leave out the "expand" bit, then you get bucket indexes rather than
values.  For a linear bucketizer of step 10, bucket index 2 covers ranges 20 to
30.  This format is generally more useful when you're shipping data around or
doing other aggregations or calculations with it.

You can break out these results by host (and using the un-expanded form):

```javascript
assert.deepEqual(
    skinner.aggregate(datapoints, [ 'host', 'util' ], bucketizers),
    [ [ 'host1', 1, 1 ],
      [ 'host1', 8, 1 ],
      [ 'host2', 3, 1 ],
      [ 'host2', 5, 1 ],
      [ 'host3', 0, 1 ],
      [ 'host3', 8, 1 ],
      [ 'host4', 0, 1 ],
      [ 'host4', 9, 1 ] ]);

```

Besides the linear bucketizer, there's a log-linear bucketizer.  For details on
what that does, see the DTrace llquantize() function.  To see how to use it,
check the source for `makeLogLinearBucketizer`.


## Streaming interface

For large numbers of data points where you don't want to keep all data points in
memory at once, you can use the object-mode streaming interface.  You write JSON
objects to it, and it emits the final summary.  You can also fetch the summary
so far at any time using the result() method.  For example, using the city/state
datapoints above:

```javascript
bucketizers = {};

stream = skinner.createAggregator({
    'bucketizers': bucketizers,
    'decomps': [ 'city' ]
});

datapoints.forEach(function (pt) { stream.write(pt); });
stream.end();

/* These two print the same thing. */
console.log(stream.result());
stream.on('data', function (result) { console.log(result); });
```


## Notes

Error checking is not great at the moment.  (Most input errors result in
assertion failures.)  Patches welcome.

You might also want to check out [krill](http://github.com/joyent/node-krill),
which filters similar-looking data points.  Between krill and skinner, you can
slice and dice data points in lots of different ways.

## Test

To test this library, run the 'test' target:

```
make test
```

## Contributing

See the separate [contribution guidelines](CONTRIBUTING.md).


CHANGES.md
# Changelog

## Not yet released
* #8 skAggregator.result() should honor resultsAsPoints
* #9 skinner repo needs some TLC


tests/tst.basic.js
/*
 * Copyright (c) 2017, Joyent, Inc.
 */

var mod_assertplus = require('assert-plus');
var mod_path = require('path');
var mod_skinner = require('../lib/skinner');
var mod_vasync = require('vasync');
var datapoints, bucketizers;

/*
 * Skip style-checking on this whole file because it's more useful in this
 * particular case to have manually laid-out data structures that don't quite
 * line up with jsstyle.
 */
/* BEGIN JSSTYLED */

/*
 * Example data points: populations of US cities
 */
datapoints = [
    { 'fields': { 'city': 'Springfield', 'state': 'MA' }, 'value': 153000 },
    { 'fields': { 'city': 'Boston',      'state': 'MA' }, 'value': 636000 },
    { 'fields': { 'city': 'Worcestor',   'state': 'MA' }, 'value': 183000 },
    { 'fields': { 'city': 'Fresno',      'state': 'CA' }, 'value': 505000 },
    { 'fields': { 'city': 'Springfield', 'state': 'OR' }, 'value':  60000 },
    { 'fields': { 'city': 'Portland',    'state': 'OR' }, 'value': 600000 }
];

/* Check the sum of all populations. */
mod_assertplus.deepEqual(mod_skinner.aggregate(datapoints), [ 2137000 ]);

/* Check the sums of populations by state. */
mod_assertplus.deepEqual(mod_skinner.aggregate(datapoints, [ 'state' ]),
    [ [ 'MA', 972000 ],
      [ 'CA', 505000 ],
      [ 'OR', 660000 ] ]);

/* Check the sums of populations, broken out by city name (NOT state) */
mod_assertplus.deepEqual(mod_skinner.aggregate(datapoints, [ 'city' ]),
    [ [ 'Springfield', 213000 ],
      [ 'Boston',      636000 ],
      [ 'Worcestor',   183000 ],
      [ 'Fresno',      505000 ],
      [ 'Portland',    600000 ] ]);

/*
 * Check the sums of populations, broken out by state *and* city
 * (same as the original dataset, in this case).
 */
mod_assertplus.deepEqual(mod_skinner.aggregate(datapoints, [ 'state', 'city' ]),
    [ [ 'MA', 'Springfield', 153000 ],
      [ 'MA', 'Boston',      636000 ],
      [ 'MA', 'Worcestor',   183000 ],
      [ 'CA', 'Fresno',      505000 ],
      [ 'OR', 'Springfield', 60000 ],
      [ 'OR', 'Portland',    600000 ] ]);


/*
 * Test nested properties, which should use pluck().
 */
datapoints = [
    { 'fields': { 'data': { 'city': 'Springfield', 'state': 'MA' } }, 'value': 153000 },
    { 'fields': { 'data': { 'city': 'Boston',      'state': 'MA' } }, 'value': 636000 },
    { 'fields': { 'data': { 'city': 'Worcestor',   'state': 'MA' } }, 'value': 183000 },
    { 'fields': { 'data': { 'city': 'Fresno',      'state': 'CA' } }, 'value': 505000 },
    { 'fields': { 'data': { 'city': 'Springfield', 'state': 'OR' } }, 'value':  60000 },
    { 'fields': { 'data': { 'city': 'Portland',    'state': 'OR' } }, 'value': 600000 }
];
mod_assertplus.deepEqual(mod_skinner.aggregate(datapoints, [ 'data.state' ]),
    [ [ 'MA', 972000 ],
      [ 'CA', 505000 ],
      [ 'OR', 660000 ] ]);

/*
 * Test bucketizers.
 */
datapoints = [
    { 'fields': { 'host': 'host1', 'cpu': 'cpu0', 'util': 83 }, 'value': 1 },
    { 'fields': { 'host': 'host1', 'cpu': 'cpu1', 'util': 13 }, 'value': 1 },
    { 'fields': { 'host': 'host2', 'cpu': 'cpu0', 'util': 37 }, 'value': 1 },
    { 'fields': { 'host': 'host2', 'cpu': 'cpu1', 'util': 53 }, 'value': 1 },
    { 'fields': { 'host': 'host3', 'cpu': 'cpu0', 'util': 88 }, 'value': 1 },
    { 'fields': { 'host': 'host3', 'cpu': 'cpu1', 'util':  9 }, 'value': 1 },
    { 'fields': { 'host': 'host4', 'cpu': 'cpu0', 'util': 98 }, 'value': 1 },
    { 'fields': { 'host': 'host4', 'cpu': 'cpu1', 'util':  5 }, 'value': 1 }
];

bucketizers = {
    'util': mod_skinner.makeLinearBucketizer(10)
};

/*
 * Summarize overall CPU utilization in a single histogram.
 */
var expand = mod_skinner.ordinalToBounds.bind(null, bucketizers.util);
mod_assertplus.deepEqual(
    expand(mod_skinner.aggregate(datapoints, [ 'util' ], bucketizers)),
    [ [ [  0,  9 ], 2 ],
      [ [ 10, 19 ], 1 ],
      [ [ 30, 39 ], 1 ],
      [ [ 50, 59 ], 1 ],
      [ [ 80, 89 ], 2 ],
      [ [ 90, 99 ], 1 ] ]);

/*
 * Check CPU utilization histograms for each host.
 */
mod_assertplus.deepEqual(
    mod_skinner.aggregate(datapoints, [ 'host', 'util' ], bucketizers),
    [ [ 'host1', 1, 1 ],
      [ 'host1', 8, 1 ],
      [ 'host2', 3, 1 ],
      [ 'host2', 5, 1 ],
      [ 'host3', 0, 1 ],
      [ 'host3', 8, 1 ],
      [ 'host4', 0, 1 ],
      [ 'host4', 9, 1 ] ]);

/*
 * Check CPU utilization histograms for each CPU name.
 */
mod_assertplus.deepEqual(
    mod_skinner.aggregate(datapoints, [ 'cpu', 'util' ], bucketizers),
    [ [ 'cpu0', 3, 1 ],
      [ 'cpu0', 8, 2 ],
      [ 'cpu0', 9, 1 ],
      [ 'cpu1', 0, 2 ],
      [ 'cpu1', 1, 1 ],
      [ 'cpu1', 5, 1 ] ]);

/*
 * Test streaming aggregators.
 */
datapoints = [
	{ fields: { 'test': 'basic', 'aggregatedField': 1 }, value: 1},
	{ fields: { 'test': 'advanced', 'anotherAggregatedField': 2 }, value: 2}
];

/*
 * Streaming aggregator with results as a sum.
 */
function streamResultsAsSum(callback) {
	var stream = mod_skinner.createAggregator({
		decomps: ['test']
	});
	datapoints.forEach(function (pt) { stream.write(pt); });
	stream.end();

	stream.on('data', function (_) {
		/*
		 * We'll ignore the 'result' argument that this callback
		 * provides since we're only using this callback as a way to
		 * ensure all of the points have been aggregated.
		 */
		mod_assertplus.deepEqual(stream.result(),
		    [['basic', 1], ['advanced', 2]]);
		callback();
	});
}

/*
 * Streaming aggregator with results as datapoints.
 */
function streamResultsAsDatapoints(callback) {
	var stream = mod_skinner.createAggregator({
		decomps: ['test'],
		resultsAsPoints: true
	});
	datapoints.forEach(function (pt) { stream.write(pt); });
	stream.end();

	stream.on('data', function (_) {
		mod_assertplus.deepEqual(stream.result(), [
		    { 'fields': { 'test': 'basic' }, 'value': 1},
		    { 'fields': { 'test': 'advanced' }, 'value': 2}
		]);
		callback();
	});
}

mod_vasync.parallel({
    'funcs': [
	streamResultsAsSum,
	streamResultsAsDatapoints
    ]
}, function (err, result) {
	var file = mod_path.basename(process.argv[1]);
	if (err) {
		console.error('%s: error running tests: %s', file, err.message);
		process.exit(1);
	}
	console.log('test %s okay', file);
});

/* END JSSTYLED */


tests/tst.nonnumeric.js
/*
 * Copyright (c) 2017, Joyent, Inc.
 */

var mod_assertplus = require('assert');
var mod_path = require('path');
var mod_skinner = require('../lib/skinner');

var datapoints, bucketizers, stream;
var gotwarning = false, gotdata = false;

datapoints = [
    { 'fields': { 'city': 'Boston',      'pop': 636000 }, 'value': 1 },
    { 'fields': { 'city': 'Worcestor',   'pop': 'bogus!' }, 'value': 1 },
    { 'fields': { 'city': 'Springfield', 'pop': 153000 }, 'value': 1 }
];
bucketizers = {
    'pop': mod_skinner.makeLinearBucketizer(100000)
};
stream = mod_skinner.createAggregator({
    'decomps': [ 'pop' ],
    'bucketizers': bucketizers
});
stream.on('invalid_object', function (obj, err, num) {
	mod_assertplus.deepEqual(obj, {
	    'fields': {
		'city': 'Worcestor',
		'pop': 'bogus!'
	    },
	    'value': 1
	});
	mod_assertplus.equal(num, 2);
	mod_assertplus.equal(err.message,
	    'value for field "pop" is not a number');
	gotwarning = true;
});
datapoints.forEach(function (d) { stream.write(d); });
stream.end();
stream.on('data', function (results) {
	mod_assertplus.deepEqual(results, [ [ 1, 1 ], [ 6, 1 ] ]);
	gotdata = true;
});

stream.on('end', function () {
	console.log('test %s okay', mod_path.basename(process.argv[1]));
	mod_assertplus.ok(gotdata);
	mod_assertplus.ok(gotwarning);
});


tests/tst.bucketize_p2.js
/*
 * Copyright (c) 2017, Joyent, Inc.
 */

var mod_assertplus = require('assert');
var mod_path = require('path');
var mod_skinner = require('../lib/skinner');

var bucketizer = mod_skinner.makeP2Bucketizer();
var func = bucketizer.bucketize.bind(bucketizer);
var expand = mod_skinner.ordinalToBounds.bind(null, bucketizer);
var dist = [];
var i;

/*
 * Test obvious special cases near the bottom of the range.
 */
func(dist, 1, 3);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 1, 1 ], 3 ]
]);

func(dist, 0, 5);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0, 0 ], 5 ],
    [ [ 1, 1 ], 3 ]
]);

func(dist, 1, 7);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0, 0 ], 5 ],
    [ [ 1, 1 ], 10 ]
]);

func(dist, 0, 2);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0, 0 ], 7 ],
    [ [ 1, 1 ], 10 ]
]);

func(dist, 2, 9);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0, 0 ], 7 ],
    [ [ 1, 1 ], 10 ],
    [ [ 2, 3 ], 9 ]
]);

func(dist, 3, 3);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0, 0 ], 7 ],
    [ [ 1, 1 ], 10 ],
    [ [ 2, 3 ], 12 ]
]);

func(dist, 4, 1);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0, 0 ], 7 ],
    [ [ 1, 1 ], 10 ],
    [ [ 2, 3 ], 12 ],
    [ [ 4, 7 ], 1 ]
]);

/*
 * Test filling in sparse areas.
 */
dist = [];
func(dist, 9, 1);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 8, 15 ], 1 ]
]);

func(dist, 3, 7);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 2, 3 ], 7 ],
    [ [ 8, 15 ], 1 ]
]);

func(dist, 4, 2);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 2, 3 ], 7 ],
    [ [ 4, 7 ], 2 ],
    [ [ 8, 15 ], 1 ]
]);

/* Test everything up to 31. */
dist = [];
for (i = 0; i < 32; i++)
	func(dist, i, 1);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0,  0],  1 ],
    [ [ 1,  1],  1 ],
    [ [ 2,  3],  2 ],
    [ [ 4,  7],  4 ],
    [ [ 8, 15],  8 ],
    [ [16, 31], 16 ]
]);

/* Test *adding* someting to each of those. */
for (i = 0; i < 32; i++)
	func(dist, i, 2);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0,  0],  3 ],
    [ [ 1,  1],  3 ],
    [ [ 2,  3],  6 ],
    [ [ 4,  7], 12 ],
    [ [ 8, 15], 24 ],
    [ [16, 31], 48 ]
]);

/* Test the same thing backwards. */
dist = [];
for (i = 31; i >= 0; i--)
	func(dist, i, 3);
mod_assertplus.deepEqual(expand(dist), [
    [ [ 0,  0],  3 ],
    [ [ 1,  1],  3 ],
    [ [ 2,  3],  6 ],
    [ [ 4,  7], 12 ],
    [ [ 8, 15], 24 ],
    [ [16, 31], 48 ]
]);
console.log('test %s okay', mod_path.basename(process.argv[1]));


examples/streaming.js
/*
 * examples/streaming.js: outline streaming interface
 */

var util = require('util');
var skinner = require('../lib/skinner');
var datapoints, bucketizers, stream;

/*
 * See the "basic" example first.
 */
bucketizers = {};
datapoints = [
    { 'fields': { 'city': 'Springfield', 'state': 'MA' }, 'value': 153000 },
    { 'fields': { 'city': 'Boston',      'state': 'MA' }, 'value': 636000 },
    { 'fields': { 'city': 'Worcestor',   'state': 'MA' }, 'value': 183000 },
    { 'fields': { 'city': 'Fresno',      'state': 'CA' }, 'value': 505000 },
    { 'fields': { 'city': 'Springfield', 'state': 'OR' }, 'value':  60000 },
    { 'fields': { 'city': 'Portland',    'state': 'OR' }, 'value': 600000 }
];

stream = skinner.createAggregator({
    'bucketizers': bucketizers,
    'decomps': [ 'city' ]
});

datapoints.forEach(function (pt) { stream.write(pt); });
stream.end();

/* These two print the same thing. */
console.log(stream.result());
stream.on('data', function (result) { console.log(result); });


examples/basic.js
/*
 * examples/basic.js: outline basic usage
 */

var util = require('util');
var skinner = require('../lib/skinner');
var datapoints, bucketizers;

/*
 * Example data points: populations of US cities
 */
datapoints = [
    { 'fields': { 'city': 'Springfield', 'state': 'MA' }, 'value': 153000 },
    { 'fields': { 'city': 'Boston',      'state': 'MA' }, 'value': 636000 },
    { 'fields': { 'city': 'Worcestor',   'state': 'MA' }, 'value': 183000 },
    { 'fields': { 'city': 'Fresno',      'state': 'CA' }, 'value': 505000 },
    { 'fields': { 'city': 'Springfield', 'state': 'OR' }, 'value':  60000 },
    { 'fields': { 'city': 'Portland',    'state': 'OR' }, 'value': 600000 }
];

/* Print the sum of all populations. */
console.log(skinner.aggregate(datapoints));

/* Print the sums of populations, broken out by state. */
console.log(skinner.aggregate(datapoints, [ 'state' ]));

/* Print the sums of populations, broken out by city name (NOT state) */
console.log(skinner.aggregate(datapoints, [ 'city' ]));

/*
 * Print the sums of populations, broken out by state *and* city
 * (same as the original dataset in this case).
 */
console.log(skinner.aggregate(datapoints, [ 'state', 'city' ]));

/*
 * When one of the fields is numeric, it may not make sense to break out all
 * discrete values, but to group them into buckets (i.e., a histogram).  Here's
 * a set of data points describing CPU utilization on four 2-CPU systems.  (As
 * above, the "value" here is the number of CPUs from host "host" with cpu name
 * "cpu" having utilization "util".  It's always 1 in this case.)
 */
datapoints = [
    { 'fields': { 'host': 'host1', 'cpu': 'cpu0', 'util': 83 }, 'value': 1 },
    { 'fields': { 'host': 'host1', 'cpu': 'cpu1', 'util': 13 }, 'value': 1 },
    { 'fields': { 'host': 'host2', 'cpu': 'cpu0', 'util': 37 }, 'value': 1 },
    { 'fields': { 'host': 'host2', 'cpu': 'cpu1', 'util': 53 }, 'value': 1 },
    { 'fields': { 'host': 'host3', 'cpu': 'cpu0', 'util': 88 }, 'value': 1 },
    { 'fields': { 'host': 'host3', 'cpu': 'cpu1', 'util':  9 }, 'value': 1 },
    { 'fields': { 'host': 'host4', 'cpu': 'cpu0', 'util': 98 }, 'value': 1 },
    { 'fields': { 'host': 'host4', 'cpu': 'cpu1', 'util':  5 }, 'value': 1 }
];

/*
 * We'll define a linear bucketizer with step "10", which means we'll group the
 * "util" values into equal-sized buckets of 10 units each.
 */
bucketizers = {
    'util': skinner.makeLinearBucketizer(10)
};

/*
 * Summarize overall CPU utilization in a single histogram.
 */
console.log(util.inspect(
    skinner.aggregate(datapoints, [ 'util' ], bucketizers), false, 5));

/*
 * Print a CPU utilization histogram *for each host*.
 */
console.log(util.inspect(
    skinner.aggregate(datapoints, [ 'host', 'util' ], bucketizers), false, 5));

/*
 * Print a CPU utilization histogram *for each CPU* (combining utilization from
 * different hosts into the same histogram).
 */
console.log(util.inspect(
    skinner.aggregate(datapoints, [ 'cpu', 'util' ], bucketizers), false, 5));
